Serialize and deserialize methods

MummenRider opened the pull request:
I decided to create a PR for this because I changed a lot of files already.
As you can see, a lot of boilerplate codes. I tried to use a plugin for this as per Flutter documentation, but won't let me because of nullable (requires me to upgrade)

To visualize our data, creating a json like this helps me to understand the structure.
This structure is based in our current models.

Eagle view
```
{
  "atrId": 123,
  "shipperInfo": {.....},
  "TransporterInfo": {.....},
  "FeedWaterRestInfo": {.....},
  "LoadingVehicleInfo": {.....},
  "DeliveryInfo": {.....},
  "AcknowledgementInfo": {.....},
  "ContingencyPlanInfo": {.....},
}
```
Expanded view
```
{
  "atrId": "1235",
  "shipperInfo": {
    "shipperName": "Clark",
    "shipperIsAnimalOwner": false,
    "departureLocationId": "123",
    "departureLocationName": "SomePlace",
    "shipperContactInfo": "213",
    "departureAddress": {
      "street": "Bruce",
      "city": "Regina",
      "postalCode": "dfsdf",
      "country": "Canada"
    }
  },
  "transporterInfo": {
    "companyName": "Company A",
    "companyAddress": {
      "street": "Bruce",
      "city": "Regina",
      "postalCode": "dfsdf",
      "country": "Canada"
    },
    "_driverNames": [
      "Name 1",
      "Name 2"
    ],
    "vehicleProvince": "vehicleProvince Name",
    "vehicleLicensePlate": "vehicleLicensePlate plate",
    "trailerProvince": "trailerProvince province",
    "trailerLicensePlate": "trailerLicensePlate plate",
    "dateLastCleaned": "this is date",
    "addressLastCleanedAt": {
      "street": "Bruce",
      "city": "Regina",
      "postalCode": "dfsdf",
      "country": "Canada"
    },
    "driversAreBriefed": false,
    "driversHaveTraining": false,
    "trainingType": "trainingType type",
    "trainingExpiryDate": "this is date"
  },
  "FeedWaterRestInfo": {
    "lastFwrDate": "This is date",
    "lastFwrLocation": {
      "street": "Bruce",
      "city": "Regina",
      "postalCode": "dfsdf",
      "country": "Canada"
    },
    "feedWaterRestEvent": [
      {
        "animalsWereUnloaded": false,
        "fwrTime": "this is date",
        "lastFwrLocation": {
          "street": "Bruce",
          "city": "Regina",
          "postalCode": "dfsdf",
          "country": "Canada"
        },
        "fwrProvidedOnboard": false
      }
    ]
  },
  "loadingVehicleInfo": {
    "dateAndTimeLoaded": "This is a date",
    "loadingArea": 1,
    "loadingDensity": 2,
    "animalsPerLoadingArea": 3,
    "animalGroup": [
      {
        "species": "species A",
        "groupAge": 1,
        "approximateWeight": 2,
        "animalPurpose": "hello",
        "numberAnimals": 2,
        "animalsFitForTransport": false,
        "_compromisedAnimal": [
          {
            "animalDescription": "animalDescription",
            "measuresTakenToCareForAnimal": "measuresTakenToCareForAnimal"
          }
        ],
        "_specialNeedsAnimal": [
          {
            "animalDescription": "animalDescription",
            "measuresTakenToCareForAnimal": "measuresTakenToCareForAnimal"
          }
        ]
      }
    ]
  },
  "deliveryInfo": {
    "receiverInfo": {
      "receiverCompanyName": "receiverCompanyName a",
      "receiverName": "receiverName A",
      "accountId": "accountId",
      "destinationLocationId": "destinationLocationId a",
      "destinationLocationName": "destinationLocationName a",
      "destinationAddress": {
        "street": "Bruce",
        "city": "Regina",
        "postalCode": "dfsdf",
        "country": "Canada"
      },
      "receiverContactInfo": "receiverContactInfo"
    },
    "arrivalDateAndTime": "this is a date",
    "compromisedAnimals": [
      {
        "animalDescription": "animalDescription",
        "measuresTakenToCareForAnimal": "measuresTakenToCareForAnimal"
      }
    ],
    "additionalWelfareConcerns": "additionalWelfareConcerns"
  },
  "acknowledgementInfo": {
    "shipperAck": "file",
    "transporterAck": "file",
    "receiverAck": "file"
  },
  "contingencyPlanInfo": {
    "goalStatement": "goalStatement",
    "communicationPlan": "communicationPlan",
    "_crisisContacts": [
      "a",
      "b"
    ],
    "expectedPrepProcess": "expectedPrepProcess",
    "standardAnimalMonitoring": "standardAnimalMonitoring",
    "_potentialHazards": [
      "a",
      "b"
    ],
    "_potentialSafetyActions": [
      "a",
      "b"
    ],
    "_contingencyEvents": [
      {
        "eventDateAndTime": "this is a time",
        "_producerContactsUsed": [
          "a",
          "b"
        ],
        "_receiverContactsUsed": [
          "a",
          "b"
        ],
        "disturbancesIdentified": "disturbancesIdentified",
        "_activities": [
          {
            "time": "this is a time",
            "personContacted": "personContacted",
            "methodOfContact": "methodOfContact",
            "instructionsGiven": "instructionsGiven"
          }
        ]
      }
    ]
  }
}
```

Note: Some field names do not match in our current models. I will change the field names and will be located under test_data for testing purposes

MummenRider committed:
added toJson and fromJson to each models

MummenRider commented:
Found current issue in this PR during testing: Will fix DateTime and field names

MummenRider committed:
fixed formatting issues in fromJson

holtzmak commented on app/lib/core/models/animal_transport_record.dart:
```diff
@@ -17,15 +17,43 @@ class AnimalTransportRecord {
   final DeliveryInfo deliveryInfo;
   final AcknowledgementInfo ackInfo;
   final ContingencyPlanInfo contingencyInfo;
+  final String atrId;
```
- [x] Shall we add the `isComplete` field as well?
The ATR ID is the uniquely identifying ID from firebase right, when we make an ATR, we should set the ID. How can we do this if the ATR is made before it's inserted into the database?

holtzmak replied:
Maybe also add `ATRwithID` `ATRwithUserId` methods.

MummenRider replied:
- I will add isComplete field in `AnimalTransportRecord` in my next commit
- The userId's purpose is to check who created that specific record. It uniquely identifies who created the record. 
- atrId's purpose is to uniquely identify each atr record

> when we make an ATR, we should set the ID 

This task will be covered in atr newRecord query. The atrId will be set before inserting the new record.

- Where are we going to use the suggested methods?

holtzmak replied:
Ah, right. I got ahead of myself. We won't need to overwrite the ATR or User identifier because these don't change.

- [ ] We should add a `withMarkedComplete()` method to set the isComplete field.

_Recall this ATR is immutable, so we should not make `isComplete` not final to avoid needing `withMarkedComplete()`. We should use `with()`._

holtzmak commented on app/lib/core/models/contingency_plan_info.dart:
```diff
@@ -26,6 +26,32 @@ class ContingencyPlanInfo {
+  Map<String, dynamic> toJSON() => {
+        'goalStatement': goalStatement,
+        'communicationPlan': communicationPlan,
+        '_crisisContacts': _crisisContacts,
```
- [x] No underscore, the underscore is just to define private methods in dart files. So, not necessary here.
We can also do this later, not a big deal.

holtzmak commented on app/lib/core/models/receiver_info.dart:
```diff
@@ -22,6 +22,25 @@ class ReceiverInfo {
+  Map<String, dynamic> toJSON() => {
+        'receiverCompanyName': receiverCompanyName,
+        'receiverName': receiverName,
+        'accountId': receiverCompanyName ?? Optional.empty(),
```
Should not be company name for account ID
```suggestion
        'accountId': accountId ?? Optional.empty(),
```

holtzmak replied:
Does the optional translate to JSON? It might need to have toJSON as well...

MummenRider replied:
Pretty sure no need to add toJSON here. I tried it and it complained

holtzmak replied:
The `Optional` type isn't a JSON primitive, I think we need to do something besides give it optional.

holtzmak replied:
My apologies, I forgot the JSON basics. https://stackoverflow.com/questions/9619852/what-is-the-convention-in-json-for-empty-vs-null.

- [x] Null is fine, but what we should do here is:
`accountID: acountID.isPresent() ? acccountID.get() : null`

holtzmak reviewed (changes requested):
Yeah, I've definitely had the misfortune of having to self-define toJSON and fromJSON methods myself. Lotta work, thank you for this.

I noticed some inconsistencies and made comments.

Also, do you think we should just bite the bullet and use `null-safety` despite beta? Or should we stick with what we have.

MummenRider committed:
Update app/lib/core/models/receiver_info.dart

thanks for this. Too much fields to edit, I was pretty sure Im going to make mistake lol.

Co-authored-by: Kelly Holtzman <32527219+holtzmak@users.noreply.github.com>

MummenRider committed:
Added isComplete record and updated test data

holtzmak commented on app/lib/core/models/receiver_info.dart:
```diff
@@ -22,6 +22,25 @@ class ReceiverInfo {
+  ReceiverInfo.fromJson(Map<String, dynamic> json)
+      : receiverCompanyName = json['receiverCompanyName'],
+        receiverName = json['receiverName'],
+        accountId = json['accountId'] ?? Optional.empty(),
```
- [x] Additionally, this should be:
`accountID = Optional.of(json[`accountID`]`

holtzmak reviewed (approved):
Minor suggestions as per your responses to my previous comments. LGTM once these are addressed.

MummenRider committed:
Changed field accountId
